1054 fix sum implementation per cql specs

igajurelNCQA opened the pull request:
Fixes https://github.com/FirelyTeam/firely-cql-sdk/issues/1054

Included submodule PRs:
* https://github.com/FirelyTeam/Ncqa.DQIC/pull/38
* https://github.com/FirelyTeam/Firely.Cql.Sdk.Integration.Runner/pull/79

Bug details:

Currently the aggregation operation for Sum does not follow CQL specs for integer, long and decimal types
Per CQL specs https://cql.hl7.org/09-b-cqlreference.html#sum
When all elements inside the input list are null or the list itself is a null list, then engine needs to return null, instead of attempting to Sum it.

Problem in Code:

See https://github.com/FirelyTeam/firely-cql-sdk/blob/develop/Cql/Cql.Runtime/Operators/CqlOperators.AggregateFunctions.cs


        public int? Sum(IEnumerable<int?> values) => values?.Sum();
        public long? Sum(IEnumerable<long?> values) => values?.Sum();
        public decimal? Sum(IEnumerable<decimal?> values) => values?.Sum();

igajurelNCQA committed:
fixes

igajurelNCQA committed:
sobmodule update

igajurelNCQA committed:
Revert "sobmodule update"

This reverts commit 8c4b444b1e1d066352e9da67e4c0c7dff68443a1.

baseTwo committed:
submodules

baseTwo committed:
Refactor Sum methods with CqlSum extension

Introduce a new static class `CqlMath` with a generic extension method `CqlSum` to calculate the sum of nullable numeric values in an `IEnumerable`. This method leverages the `IAdditionOperators` interface for type flexibility.

Refactor `Sum` methods in `CqlOperators.AggregateFunctions.cs` to use `CqlSum`, simplifying code by removing manual null-checking and summation logic. Update method signatures to accept nullable `IEnumerable` parameters.

Modify `Sum` method for `CqlQuantity` to use pattern matching for null-checking. Update `ICqlOperators.cs` interface to reflect these changes.

Note a minor change in the `Ncqa.DQIC` subproject indicating a dirty state.

baseTwo committed:
Enhance Sum methods with error handling

Updated Sum methods in CqlOperators.AggregateFunctions.cs to include try-catch blocks for handling OverflowException and NotSupportedException, logging warnings and returning null on errors. Refactored CqlSum in CqlMath to use INumberBase<TNumber> for a more general numeric interface and added a new CqlAggregate method for flexible aggregation. Marked Ncqa.DQIC subproject as dirty due to uncommitted changes.

baseTwo committed:
Refactor CqlSum and CqlAggregate methods

Refactored the CqlSum method to utilize the CqlAggregate method, reducing code duplication and improving clarity. Updated the CqlAggregate method signature to use nullable types for the aggregator function and initialAccumulate parameter, enhancing flexibility. Simplified a return statement in a switch case by using the default keyword. These changes improve code maintainability and readability.

baseTwo commented:
@igajurelNCQA , All but the NCQA tests are passing. These have to be fixed by someone on your side. I know @richfirely  got them from NCQA some time ago, maybe they need to be updated. cc @alexzautke

igajurelNCQA commented:
> @igajurelNCQA , All but the NCQA tests are passing. These have to be fixed by someone on your side. I know @richfirely got them from NCQA some time ago, maybe they need to be updated. cc @alexzautke

@baseTwo Tests were passing yesterday against my dqic branch 1054-Hedis-CQL-fixes-to-align-with-Sum-engine-fixes. Let me retest and also repull to get latest on this branch and retest and let you know.

baseTwo committed:
Allow nullable enumerables in Sum method signatures

Updated the `PublicAPI.Shipped.txt` file to modify the `Sum` method signatures in the `Hl7.Cql.Operators.ICqlOperators` interface, allowing nullable enumerables as input parameters. This change affects the `Sum` methods for `decimal`, `Hl7.Cql.Primitives.CqlQuantity`, `int`, and `long` types. Additionally, enabled nullable reference types in `PublicAPI.Unshipped.txt` by adding `#nullable enable`.

baseTwo committed:
Update Ncqa.DQIC submodule to new commit

The subproject commit for `Ncqa.DQIC` has been updated from `8434f403cffd168fe63ad64b1c1474eb242ebed9` to `5258fc2f1189ac43062a23af153742dc20dfbf9c`. This update may include new features, bug fixes, or other changes in the subproject's repository.

baseTwo commented:
> > @igajurelNCQA , All but the NCQA tests are passing. These have to be fixed by someone on your side. I know @richfirely got them from NCQA some time ago, maybe they need to be updated. cc @alexzautke
> 
> @baseTwo Tests were passing yesterday against my dqic branch 1054-Hedis-CQL-fixes-to-align-with-Sum-engine-fixes. Let me retest and also repull to get latest on this branch and retest and let you know.

@igajurelNCQA I changed the NCQA branch to the one you mentioned. The tests are still failing as before. Could you have a look?

igajurelNCQA committed:
subdmodule reference
